src/plugins/intel_gpu/src/kernel_selector/kernels/mvn/mvn_kernel_bfyx_opt.cpp

```diff
@@ -37,52 +38,89 @@ MVNKernelBfyxOpt::Parent::DispatchData MVNKernelBfyxOpt::SetDefault(const mvn_pa
-        dispatchData.lws[0] *= 2;
-        dispatchData.itemsNum /= 2;
-    }
-
-    dispatchData.gws[0] = dispatchData.lws[0];
-    dispatchData.leftovers = dispatchData.dataSetSize % dispatchData.lws[0];
+        // start with 1 thread per data set
+        dispatchData.gws[0] = 1;
+        dispatchData.gws[1] = dispatchData.dataSetsCount;
+        dispatchData.gws[2] = 1;
+        dispatchData.itemsNum = dispatchData.dataSetSize;
+
+        dispatchData.lws[0] = 1;
+        dispatchData.lws[1] = 1;
+        dispatchData.lws[2] = 1;
+        // Compute maximum possible LWS that does not exceed device capabilities and optimizes number of global memory
+        // reads.
+        while ((dispatchData.itemsNum > 32 || dispatchData.lws[0] < dispatchData.itemsNum) && (2 * dispatchData.lws[0] <= max_lws)) {
+            dispatchData.lws[0] *= 2;
+            dispatchData.itemsNum /= 2;
+        }
 
+        dispatchData.gws[0] = dispatchData.lws[0];
+        dispatchData.leftovers = dispatchData.dataSetSize % dispatchData.lws[0];
+    }
     return dispatchData;
 }
 
 JitConstants MVNKernelBfyxOpt::GetJitConstants(const mvn_params& params, MVNKernelBase::DispatchData dispatchData) const {
     auto jit = MVNKernelBase::GetJitConstants(params, dispatchData);
 
-    jit.AddConstants({
-        MakeJitConstant("ITEMS_NUM", dispatchData.itemsNum),
-        MakeJitConstant("LWS", dispatchData.lws[0]),
-        MakeJitConstant("GWS", dispatchData.gws[0]),
-        MakeJitConstant("DATA_SETS_COUNT", dispatchData.dataSetsCount),
-        MakeJitConstant("DATA_SET_SIZE", dispatchData.dataSetSize),
-        MakeJitConstant("LEFTOVERS", dispatchData.leftovers),
-    });
+    if (params.has_dynamic_tensors()) {
+        const auto& input = params.inputs[0];
+        auto x = toCodeString(input.X(), 5);
+        auto y = toCodeString(input.Y(), 4);
+        auto z = toCodeString(input.Z(), 3);
+        auto w = toCodeString(input.W(), 2);
+        auto f = toCodeString(input.Feature(), 1);
+        auto b = toCodeString(input.Batch(), 0);
+        std::string data_set_size;
+        std::string data_set_count;
+        if (params.mvnMode == MVNMode::WITHIN_CHANNELS) {
+            data_set_size = toVectorMulString({x, y, z});
+            data_set_count = toVectorMulString({f, b});
+        } else {
+            data_set_size = toVectorMulString({x, y, z, f});
+            data_set_count = b;
+        }
+        // since lws[0] is calculated by power of 2
+        // items_num can be calculated by dividing data_set_size by power of 2
+        const std::string lws_0 = "get_local_size(0)";
+        const std::string calc_power = "({uint pos = 0; uint i = n; do { i >>= 1; ++pos; } while (i); pos - 1;})";
+        const std::string power = "CALC_POWER(" + lws_0 + ")";
+        const std::string items_num = "(DATA_SET_SIZE>>POWER)";
+        const std::string left_overs = "(DATA_SET_SIZE-(ITEMS_NUM<<POWER))";
```

Comment:
And I think all of these can be just written in the cl code directly, no need to be created through CreateJit.

Reply:
moved compile time constants to cl code.